understanding/20/keyboard-no-exception.html

```diff
@@ -27,8 +27,35 @@ <h2>Intent of Keyboard (No Exception)</h2>
          function requires input that depends on the path of the user's movement and not just
          the endpoints (excluded from the requirements of 2.1.1) must be made keyboard accessible.
          Rather, it means that content that uses path-dependent input cannot conform to this
-         Success Criterion and therefore cannot meet Guideline 2.1 at Level AAA. 
+         Success Criterion and therefore cannot meet Guideline 2.1 at Level AAA.
       </p>
+
+      <div class="note">
+         <p>Platforms and user agents usually have conventions for how web content or
+            applications are controlled with a keyboard interface. If content does not follow
+            the platform/user agent conventions it may be difficult to use, as users will need
+            to learn different interaction methods. As a <em>best practice</em>, content
+            should follow the platform/user agent conventions. However, deviating from these
```

Comment:
If keyboard operaton works via obscure undocumented keys but operation works fine for mouse users (so a clear disability-related issue) - would it not least be required for conformance that that keyboard operation is documented in context (or in help)?

Reply:
if you can normatively and authoritatively define what "obscure" is (or what exactly the "platform conventions" are, or what "non-standard" means, or similar ... even though there is no "standard"), possibly. but because these are malleable and undefined concepts, it gets tricky. also, can't start *requiring* anything here as that would make it a normative change that goes beyond what the normative SC text says? can strongly hint at it / suggest it as a best practice, but not require it.

Reply:
I agree it is tricky - but defining "obscure" (not wedded to the term)  does not seem too difficult. It would be anything that flaunts common expectations ("affordances") of interactive elements (mostly that they can be navigated with TAB and arrow keys and activated with ENTER and/or space bar).
If someone created a scripted custom control looking like a button to respond **only** to to CTR+$ to activate, that would be keyboard accessible **if anyone knew it** - but no one would without documentation. Hence the requirement to document the obscurity. Unless you delve into scripts, an evaluation against 2.1.1 would try to activate that pseudo button with ENTER and come to the same conclusion that we have a FAIL or 2.1.1. Would anyone seriously argue that this judgment would be wrong, tha th thing is keyboard-operable, if the only way to discover such operability would be to delve ito scripts to look for event handlers? I believe not...

Reply:
again, this can be added as a strong suggestion, but this interpretation goes beyond the normative wording of the SC which does not impose any such limits. my PR here simply tries to make this fact (that the normative wording may not be as stringent as folks would like)

> All [functionality](https://www.w3.org/TR/WCAG22/#dfn-functionality) of the content is operable through a [keyboard interface](https://www.w3.org/TR/WCAG22/#dfn-keyboard-interface) without requiring specific timings for individual keystrokes, except where the underlying function requires input that depends on the path of the user's movement and not just the endpoints.

Reply:
> Would anyone seriously argue that this judgment would be wrong, that the thing is keyboard-operable and passes 2.1.1

not sure an argument of "surely nobody would try to argue they're fulfilling the exact wording of the normative SC" is reason enough to try and introduce new tighter normative requirements through the backdoor of an informative document

Reply:
Well - many SCs leave room for interpretation. Understanding texts are meant to guide authors and auditors in arriving at a sound judgment in all those cases of "it depends" (TM @patrickhlauke ?). These informative texts can never be definitive and cover all cases. Other SCs have similar wiggle room, and we can and have to live with that (@mbgower had some good points on the meaning of "descriptive" above). In my view in is not helpful to suggest that having **any kind of theoretic and undocumented keyboard operability** would be enough to meet this SC. Explaining what 'keyboard operability' is commonly supposed to mean based user expectations, and offering that as a help for authors/auditors to determine whether they meet the intention of the SC seems more useful than a bare-bones contraction to the point of 'anything will do'. It is certainly helpful to explain that normatively it is be sufficient to support just ENTER and not space bar (one of two expected behaviors), but weird undocumented keyboard commands that no one has a chance of guessing are a different kettle of fish. I strongly object to the change as it is now proposed (for whatever that's worth from an invited expert).

Reply:
> many SCs leave room for interpretation

sure, but they have wiggle room in the normative language. in this case, i don't see that ambiguity/flexibility in the normative SC text that would then warrant defining a *requirement* in the understanding part. i.e. "this fails 2.1.1 because you're using an unusual keyboard shortcut/combination without documenting it" ... where does 2.1.1's normative text hint at this "if you don't document it, it fails this SC"?

Reply:
> where does 2.1.1's normative text hint at this "if you don't document it, it fails this SC"?

I think it is implicit in the term 'operable', which to in my understanding tacitly implies adherence to common keyboard conventions. So if you completely take away the conventions, you'd have to provide some clue how on earth to operate that thing. It boils down to a philosophical argument whether naked words can ever have definitive meaning, or whether implicit context is alway present and needs to be acknowledged. I could dig out some Wittgenstein on that....

Reply:
I don't disagree that Instructions should be provided for custom controls / common controls that have modified behaviors.  But I agree that requiring documentation under 2.1.1 is a stretch of what the SC normative text states.    

It would make a lot more sense, in my opinion, to require this under 3.3.2 Labels and Instructions.  Same outcome, unorthodox keyboard functionality is required to be documented for people, but just under an SC that already calls out that instructions need to be provided to people...

Reply:
The concept  of "operable through a keyboard interface" was always meant to support assistive technology like switch input that uses the keyboard interface to activate interactive items by a simple pressing of the switch. If the narrowest reading of 2.1.1 would have to accept  **any keyboard input** to activate the common controls like links and buttons  and their custom-built lookalikes instead of ENTER/space bar, that group of AT users would be excluded (if I get it right) - another point to support the argument that the idea of "keyboard operability" implies supporting the common expected activation methods for the common elements and their lookalikes.

@scottaohara Your point is valid - lack of documentation would fail 3.3.2, but I see this as a separate issue from ensuring keyboard operability in line with common user expectations. .In my view, to satisfy 2.1.1, there would have to be some commonly expected keyboard activation behavior present, and I think this is the case inn the useful examples given (activate with ENTER if space bar fails, activate with enter on the input field when the element after the search field only accepts pointe input). So just to be clear, I do not argue against including the examples in the two notes and think they are helpful.

Reply:
common user expectations may also be contingent on the platform.  at a very basic level, macOS and Windows have differences in what keyboard accessibility looks like and users have to adjust expectations if going between platforms.  So what's 'common' can potentially be different degrees of variable.

But then there's the specially made interfaces for specific users that may have been purposefully implemented with non-standard keyboard commands, because they make sense for their audience.  For instance, a game might be created that uses specific keyboard keys to activate different "buttons", or an educational application might have similarly mapped certain keyboard keys to different buttons because it is necessary/appropriate for what they are trying to teach. The "piano" UI that is mentioned in the Orientation rule.

But taking unique interfaces out of the question, I am truly curious if have you found many instances of what it is you seem to be trying to make sure is failed under 2.1.1?  A website/product taking the time to actually make their interface with custom/customized interactive elements keyboard accessible BUT doing so in a way that bucks common expectations for the platform they're on, AND they are not making these obscure keyboard behaviors known to their users, AND they need to be told by WCAG what they've done is wrong, otherwise they won't document said information or change to more common keyboard expectations?

i know i'm laying it on a bit thick there with that last bit, but I'm really struggling with understanding the scope of the potential issue that i think you're seeing.